node/src/components/consensus/era_supervisor/era.rs

```diff
@@ -161,6 +163,75 @@ impl<I> Era<I> {
     pub(crate) fn set_paused(&mut self, paused: bool) {
         self.consensus.set_paused(paused);
     }
+
+    /// Computes rewards and returns the era report.
+    pub(crate) fn create_report(
+        &self,
+        terminal_block_data: TerminalBlockData<ClContext>,
+        timestamp: Timestamp,
+        protocol_config: &ProtocolConfig,
+    ) -> EraReport<PublicKey> {
+        let TerminalBlockData {
+            inactive_validators,
+        } = terminal_block_data;
+
+        // The total rewards are the ideal number of blocks (i.e. one per minimum-length round) in
+        // the actual duration of the era, times the BLOCK_REWARD constant, independent of how many
+        // blocks were actually created.
+        let era_duration = timestamp.saturating_diff(self.start_time);
+        let min_round_length = protocol_config.highway_config.min_round_length();
+        let number_of_min_length_rounds = era_duration / min_round_length;
+        let total_rewards: u64 = number_of_min_length_rounds
+            .checked_mul(BLOCK_REWARD)
+            .unwrap_or_else(|| {
+                // With BLOCK_REWARD set to one trillion and a round length of 1 minute, a single
+                // era would have to last 35 years before the total rewards would overflow a u64.
+                error!(%era_duration, %min_round_length, %BLOCK_REWARD, "total rewards overflow");
+                u64::MAX
+            });
+
+        if total_rewards == 0 {
+            return EraReport {
+                rewards: BTreeMap::new(),
+                equivocators: self.accusations(),
+                inactive_validators,
+            };
+        }
+
+        let is_active = |public_key: &PublicKey| {
+            !(self.accusations.contains(public_key)
+                || self.faulty.contains(public_key)
+                || inactive_validators.contains(public_key))
+        };
+
+        // The total rewards are distributed proportionally by weight. To make sure that the
+        // computation doesn't overflow we divide all weights by 2^64 if the numbers are too large.
+        let total_weight: U512 = self
+            .validators
+            .iter()
+            .filter_map(|(public_key, weight)| is_active(public_key).then(|| *weight))
+            .sum();
+        let shift = if total_weight.checked_mul(total_rewards.into()).is_none() {
```

Comment:
It would be enough to check with maximum active weight instead of the total weight, wouldn't it? Though it probably doesn't make much of a difference, anyway.

Reply:
Yes, and that's already the case, isn't it? `total_weight` is already the sum of only the _active_ validators' weight.

Reply:
I meant the maximum weight of any single active validator. We never actually multiply by the total weight, only by the weights of single validators.
But that would require finding the maximum in addition to the total, and for little gain, so I'm fine either way.

Reply:
Ah, right! Yes, I don't think it's worth the added complexity.
In practice, our `U512`s won't ever overflow here. We won't even be using 128 of their bits.